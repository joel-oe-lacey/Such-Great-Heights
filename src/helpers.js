export const fetchData = () => {
  return fetch('http://localhost:3001/api/v1/areas')
    .then(response => response.json())
    .then(areasData => fetchAreasData(areasData))
}

export function fetchAreasData(areasData) {
  const promises = areasData.areas.map(area => {
    return fetch(`http://localhost:3001${area.details}`)
      .then(response => response.json())
      .then(info => {
        return {
          id: info.id,
          shortName: info.area,
          name: info.name,
          description: info.about,
          listings: info.listings,
        }
      })

  })
  return Promise.all(promises);
}

export const fetchListings = (stateAreas) => {
  const listingProms = stateAreas.reduce((areaListings, area) => {
    area.listings.forEach(listing => {
      areaListings.push(fetch(`http://localhost:3001${listing}`)
        .then(res => res.json())
        .then(info => {
          return {
            id: info.listing_id,
            area_id: info.area_id,
            name: info.name,
            details: {
              address: `${info.address.street}, ${info.address.zip}`,
              ...info.details
            }
          }
        })
      )
    })
    return areaListings
  }, []);

  return Promise.all(listingProms)
}
