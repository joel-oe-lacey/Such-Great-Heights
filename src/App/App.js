import React, { Component } from 'react';
import Login from '../Login/Login';
import Listing from '../Listing/Listing';
import Details from '../Details/Details';
import './App.scss';
import { fetchAreasData, fetchListingData } from '../helpers.js';
import Area from '../Area/Area';
import AreaContainer from '../AreaContainer/AreaContainer';
import Nav from '../Nav/Nav';
import { Route, NavLink, Switch } from 'react-router-dom'
import { Link } from 'react-router-dom'


export default class App extends Component {
  constructor() {
    super();
    this.state = {
      name: 'user1',
      tripType: 'vacation',
      areas: [],
      listings: [],
    };
  }

  componentDidMount() {
    fetch('http://localhost:3001/api/v1/areas')
      .then(response => response.json())
      .then(areasData => fetchAreasData(areasData))
      .then(areas => this.setState({ areas }))

    // fetchListingData(this.state.listingsUrl)
    //   .then(listings => this.setState({ listings }))
  }

  updateUser = userName => {
    this.setState({ name: userName })
  }

  render() {
    // let { listings } = this.state;

    /* <Details id={listings[0] ? listings[0].id : 1} name={listings[0] ? listings[0].name : 'test'} details={listings[0] ? listings[0].details : {features: []}} /> */

    // return (
    //   <section className="app">
    //     <Nav userName={this.state.name}/>
    //     <AreaContainer data={this.state.areas} tripType={this.state.tripType} />
    //   </section>
    // )

    const listings = this.state.listings.map(listing => {
      return <Listing id={listing.id} name={listing.name} />
    })

    // display login upon page load
    // upon succesful form submission, render AreasContainer
    // 'Log Out' click takes user to log in form
    // 'View Areas' click takes user to AreasContainer
    // upon 'View Listings' click display Listings component
    // upon 'View Details' click, render Details component
    return (
      <section className="app">
      <Switch>
        <Route exact path='/' component={Login} />
        <Route path='/' render={() =>  <Nav userName={this.state.name}/>} />
      </Switch>
        <Route exact path='/areas' render={() =>
          <AreaContainer data={this.state.areas} tripType={this.state.tripType} />} />

}
       </section>
    )
  }
}
