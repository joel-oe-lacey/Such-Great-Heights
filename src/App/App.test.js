import React from 'react';
import { render } from '@testing-library/react';
import App from './App.js';
import { shallow } from 'enzyme';
import { fetchAreasData, fetchListings, fetchData } from '../helpers.js';

jest.mock('../helpers.js')

describe ('App', ()=> {
  let wrapper;

  beforeEach(() => {
    fetchData.mockImplementation(() => {
    return Promise.resolve([{
    id: 590,
    name: "River North",
    location: "North of Downtown Denver",
    about: "RiNo is a burgeoning area with new bars, restaurants and event spaces popping up all the time. Explore this thriving area of Denver today!",
    region_code: 6356834,
    quick_search: "o5kod9f5cqo0",
    listings: [
        "/api/v1/listings/3",
        "/api/v1/listings/44",
        "/api/v1/listings/221",
        "/api/v1/listings/744",
        "/api/v1/listings/90",
        "/api/v1/listings/310"
    ]
}])
  });
    fetchListings.mockImplementation(() => {
  return Promise.resolve([{
  id: 590,
  name: "River North",
  location: "North of Downtown Denver",
  about: "RiNo is a burgeoning area with new bars, restaurants and event spaces popping up all the time. Explore this thriving area of Denver today!",
  region_code: 6356834,
  quick_search: "o5kod9f5cqo0",
  listings: [
      "/api/v1/listings/3",
      "/api/v1/listings/44",
      "/api/v1/listings/221",
      "/api/v1/listings/744",
      "/api/v1/listings/90",
      "/api/v1/listings/310"
  ]
}])
});
    });

  it('should add a favorite to state when addFav() is run', () => {
    wrapper = shallow(<App />);
    let mockListing = {id: 123, name: 'cool looking house'}
    let defaultState = {
      areas: [],
      favorites: [],
      listings:[],
      name: "user1",
      tripType: "vacation"
    }
    let updatedState = {
      areas: [],
      favorites: [mockListing],
      listings:[],
      name: "user1",
      tripType: "vacation"
    }
    wrapper.instance().setState(defaultState)
    wrapper.instance().addFavorite(mockListing)
    expect(wrapper.state()).toEqual(updatedState)
  })

  it('should match the snapshot', () => {
  const wrapper = shallow(<App />);

  expect(wrapper).toMatchSnapshot();
});

  it('should remove a favorite from state when removeFav is called', () => {
    wrapper = shallow(<App />);
    let mockListing = {id: 123, name: 'cool looking house'}
    let defaultState = {
      areas: [],
      favorites: [mockListing]
    }
    let updatedState = {
      areas: [],
      favorites: [],
      listings:[{
       area_id: 1,
         details: {
           features: [],
       },
           id: 1,
           name: "",
         }],
      name: "user1",
      tripType: "vacation",
    }
    wrapper.instance().setState(defaultState)
    wrapper.instance().removeFavorite(mockListing);
    expect(wrapper.state()).toEqual(updatedState);
  })

  it('should update state.name when updateUser is invoked', () => {
    wrapper = shallow(<App />);
    wrapper.instance().setState({});
    wrapper.instance().updateUser('user123');
    expect(wrapper.state()).toEqual({areas: [], favorites: [], listings: [{area_id: 1,details: {features: [], }, id: 1,name: "", }], name: 'user123', tripType: 'vacation'})
  })

  it('should retrieve ideas after mounting', () => {
    wrapper = shallow(<App />);
    expect(fetchData).toHaveBeenCalled();
  });
});
