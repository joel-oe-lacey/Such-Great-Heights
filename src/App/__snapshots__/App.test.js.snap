// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<section
  className="app"
>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      path="/"
      render={[Function]}
    />
  </Switch>
  <Route
    exact={true}
    path="/areas"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/areas/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/areas/:area_id/listings/:listing_id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
</section>
`;
